### PR TITLE
added date and history to follow + replay system

### PR DESCRIPTION
**Pull request by tsaquet**
Hello ! A small pull request to begin, to add few things : 
- date management for follows
- history for follow with a replay button, in case you missed it the first time or if it happened when not live

Once everything is ok with you i'll do the same for subs /cheers and other alerts :)
And maybe a global alert history to be able to replay all from the same place, to discuss !

I would have be pleased to add unit test results as asked, but I could not find them, tell me if i missed something or if the text is not up to date anymore.

I am not used to code with java or javascript / jquery so tell me if anything is not as expected, i will do the changes if necessary.

### CODE REVIEW

**Comment by tsaquet** [comment]
Ok, currently working on the problems found by the CI :)

**Comment by gmt2001** [comment]
By unit tests we mean just run the bot and test the changes, and show us the output relevant to the changes

**Comment by tsaquet** [comment]
Here is a screen shot showing the bot with the follow history, the replay button, the log and the alert displayed :)

image

By the way, i just found a bug, the replay is not working after the first history page, I have to find out why 🤔

File: resources/web/panel/js/pages/history/followsHistory.js [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+						toastr.success('Successfully replayed');
+					});
+				});
+			}
+		}
+	}
+
+	socket.getDBTableValues('get_all_follows_by_date', 'followedDate', function(results) {
+
+		socket.wsEvent()
```
Comment (BrandenB) [inline resources/web/panel/js/pages/history/followsHistory.js]: Improper use of the wsEvent function.

File: resources/web/panel/js/pages/history/followsHistory.js [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+			let follow = results[i];
+			let name = follow.key;
+			let date = new Date(follow.value);
+			let month = ((date.getMonth() + 1)<10? '0'+ (date.getMonth() + 1):''+(date.getMonth() + 1));
+			let day = ((date.getDate() + 1)<10? '0'+ (date.getDate() + 1):''+(date.getDate() + 1));
+
+			follows.push([
+				name,
+				date.getFullYear()+'-'+month+'-'+day,
+				'<button type="button" className="btn btn-primary pull-right" id="follows-btn-replay-'+name+'" data="'+name+'">Replay</button>'
```
Comment (BrandenB) [inline resources/web/panel/js/pages/history/followsHistory.js]: Improper way to create HTML, use jQuery. Also button should be an icon, not a name, with a tooltip like the commands tab.

Also the date should use the user's local format like all the other pages on the panel. The time should also be shown.

File: resources/web/panel/js/pages/history/followsHistory.js [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+	socket.getDBTableValues('get_all_follows_by_date', 'followedDate', function(results) {
+
+		socket.wsEvent()
+		let follows = [];
+
+		for (let i = 0; i < results.length; i++) {
+			let follow = results[i];
+			let name = follow.key;
+			let date = new Date(follow.value);
+			let month = ((date.getMonth() + 1)<10? '0'+ (date.getMonth() + 1):''+(date.getMonth() + 1));
```
Comment (BrandenB) [inline resources/web/panel/js/pages/history/followsHistory.js]: Fix code spacing.

File: resources/web/panel/js/pages/history/followsHistory.js [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+			let day = ((date.getDate() + 1)<10? '0'+ (date.getDate() + 1):''+(date.getDate() + 1));
```
Comment (BrandenB) [inline resources/web/panel/js/pages/history/followsHistory.js]: Fix code spacing.

File: resources/web/panel/js/pages/history/followsHistory.js [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+			],
+			'columns': [
+				{ 'title': 'Username', 'orderData' : [1], 'order' : [0, 'desc'] },
+				{ 'title': 'Date' },
+				{ 'title': 'Replay' },
+			]
+		});
+
+		// First time table created
+		createClickButtonEvents(results);
```
Comment (BrandenB) [inline resources/web/panel/js/pages/history/followsHistory.js]: Use table events like used in the commands tab.

File: resources/web/panel/js/pages/history/followsHistory.js [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+
+		// each time the table is drawn after page change
+		$('#followsHistoryTable').on( 'draw.dt', function () {
```
Comment (BrandenB) [inline resources/web/panel/js/pages/history/followsHistory.js]: Use table events like used in the commands tab.

File: resources/web/panel/js/pages/history/followsHistory.js [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+ */
+
+// Function that querys all of the data we need.
+$(function() {
+
+	/**
+	 *
+	 * @param results
+	 */
+	function createClickButtonEvents(results) {
```
Comment (BrandenB) [inline resources/web/panel/js/pages/history/followsHistory.js]: Use table events like used in the commands tab.

File: source/tv/phantombot/cache/FollowersCache.java [hunk 8]
```diff
@@ -122,11 +122,17 @@ private void updateCache() throws Exception {
 
                 for (int i = 0; i < jsonArray.length(); i++) {
                     String follower = jsonArray.getJSONObject(i).getJSONObject("user").getString("name").toLowerCase();
+                    String followDate = jsonArray.getJSONObject(i).getString("created_at");
 
                     if (!datastore.exists("followed", follower)) {
-                        EventBus.instance().post(new TwitchFollowEvent(follower));
+                        com.gmt2001.Console.out.println("New follower: "+follower);
```
Comment (BrandenB) [inline source/tv/phantombot/cache/FollowersCache.java]: Remove console spam

**Review by BrandenB — CHANGES_REQUESTED** [review]
This feature will brick the bot if someone has a lot of followers, if a channel uses this feature and has over 100,000 followers it will take forever for it to load, the proper way to do it would be to make it only load 100 followers at a time, similar to the leaderboard in the points section.

Also, make sure you follow the code style of the panel: 4 space indents and no random spaces left and space out the code properly.

**Comment by tsaquet** [comment]
I just rebased from the last version, starting to address the requested changes today !

File: resources/web/panel/js/pages/history/followsHistory.js (hunk 1 above)
Comment (tsaquet) [reply resources/web/panel/js/pages/history/followsHistory.js]: removed

File: resources/web/panel/js/pages/history/followsHistory.js (hunk 2 above)
Comment (tsaquet) [reply resources/web/panel/js/pages/history/followsHistory.js]: About the date format : I changed with the local format and discovered that the sorting capacity of DataTable is not able to sort french format correctly. I added a data-sort attribute with ISO format, and applied the local format to the data. Do you think there is a better way to do that ? I really need to be able to sort by date, i want the last follow to be shown first.

**Comment by tsaquet** [comment]
@BrandenB Hello ! I fixed what you asked me to do, I had to do something slightly different to be able to sort the date as I told you in the remaining requested change, tell me whether it is ok. 
I also struggled a bit to find a position for the the "load more" button, same, tell me.

**Comment by tsaquet** [comment]
Hello ! Do you have an idea when you will be able to review ? I would like to do the subs and raids, I would like to be sure the way I did the follows is ok with you before :)
Thank you !

**Comment by C74K** [comment]
Can add debug mode to check all tokens worked ??

**Comment by BrandenB** [comment]
Perfect, thanks!

**Pull request merged**